ikectl/ikectl.8
.\" $OpenBSD: ikectl.8,v 1.27 2020/04/25 18:38:21 tobhe Exp $
.\"
.\" Copyright (c) 2007-2013 Reyk Floeter <reyk@openbsd.org>
.\"
.\" Permission to use, copy, modify, and distribute this software for any
.\" purpose with or without fee is hereby granted, provided that the above
.\" copyright notice and this permission notice appear in all copies.
.\"
.\" THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
.\" WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
.\" MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
.\" ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
.\" WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
.\" ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
.\" OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
.\"
.Dd $Mdocdate: April 25 2020 $
.Dt IKECTL 8
.Os
.Sh NAME
.Nm ikectl
.Nd control the IKEv2 daemon
.Sh SYNOPSIS
.Nm
.Op Fl q
.Op Fl s Ar socket
.Ar command
.Op Ar arg ...
.Sh DESCRIPTION
The
.Nm
program controls the
.Xr iked 8
daemon and provides commands to maintain a simple X.509 certificate
authority (CA) for IKEv2 peers.
.Pp
The options are as follows:
.Bl -tag -width Ds
.It Fl q
Don't ask for confirmation of any default options.
.It Fl s Ar socket
Use
.Ar socket
instead of the default
.Pa /var/run/iked.sock
to communicate with
.Xr iked 8 .
.El
.Sh IKED CONTROL COMMANDS
The following commands are available to control
.Xr iked 8 :
.Bl -tag -width Ds
.It Cm active
Set
.Xr iked 8
to active mode.
.It Cm passive
Set
.Xr iked 8
to passive mode.
In passive mode no packets are sent to peers and no connections
are initiated by
.Xr iked 8 .
.It Cm couple
Load the negotiated security associations (SAs) and flows into the kernel.
.It Cm decouple
Unload the negotiated SAs and flows from the kernel.
This mode is only useful for testing and debugging.
.It Cm load Ar filename
Reload the configuration from the specified file.
.It Cm log brief
Disable verbose logging.
.It Cm log verbose
Enable verbose logging.
.It Cm monitor
Monitor internal messages of the
.Xr iked 8
subsystems.
.It Cm reload
Reload the configuration from the default configuration file.
.It Cm reset all
Reset the running state.
.It Cm reset ca
Reset the X.509 CA and certificate state.
.It Cm reset policy
Flush the configured policies.
.It Cm reset sa
Flush the running SAs.
.It Cm reset user
Flush the local user database.
.It Cm reset id Ar ikeid
Delete all IKE SAs with matching ID.
.It Cm show sa
Show internal state of active IKE SAs, Child SAs and IPsec flows.
.El
.Sh PKI AND CERTIFICATE AUTHORITY COMMANDS
In order to use public key based authentication with IKEv2,
a public key infrastructure (PKI) has to be set up to create and sign
the peer certificates.
.Nm
includes commands to simplify maintenance of the PKI
and to set up a simple certificate authority (CA) for
.Xr iked 8
and its peers.
.Pp
The following commands are available to control the CA:
.Bl -tag -width Ds
.It Xo
.Cm ca Ar name Cm create
.Op Cm password Ar password
.Xc
Create a new certificate authority with the specified
.Ar name .
The command will prompt for a CA password unless it is specified with
the optional
.Ar password
argument.
The password will be saved in a protected file
.Pa ikeca.passwd
in the CA directory and used for subsequent commands.
.It Cm ca Ar name Cm delete
Delete the certificate authority with the specified
.Ar name .
.It Xo
.Cm ca Ar name Cm export
.Op Cm peer Ar peer
.Op Cm password Ar password
.Xc
Export the certificate authority with the specified
.Ar name
into the current directory for transport to other systems.
This command will create a compressed tarball called
.Pa ca.tgz
in the local directory and optionally
.Pa ca.zip
if the
.Sq zip
tool is installed.
The optional
.Ar peer
argument can be used to specify the address or FQDN of the local gateway
which will be written into a text file
.Pa peer.txt
and included in the archives.
.It Xo
.Cm ca Ar name
.Cm install Op Ar path
.Xc
Install the certificate and Certificate Revocation List (CRL) for CA
.Ar name
as the currently active CA or into the specified
.Ar path .
.It Xo
.Cm ca Ar name Cm certificate Ar host
.Cm create
.Op Ic server | client | ocsp
.Xc
Create a private key and certificate for
.Ar host
and sign then with the key of certificate authority with the specified
.Ar name .
.Pp
The certificate will be valid for client and server authentication by
default by setting both flags as the extended key usage in the certificate;
this can be restricted using the optional
.Ic server
or
.Ic client
argument.
If the
.Ic ocsp
argument is specified the extended key usage will be set for OCSP signing.
.It Xo
.Cm ca Ar name Cm certificate Ar host
.Cm delete
.Xc
Deletes the private key and certificates associated with
.Ar host .
.It Xo
.Cm ca Ar name Cm certificate Ar host
.Cm export
.Op Cm peer Ar peer
.Op Cm password Ar password
.Xc
Export key files for
.Ar host
of the certificate authority with the specified
.Ar name
into the current directory for transport to other systems.
This command will create a compressed tarball
.Pa host.tgz
in the local directory and optionally
.Pa host.zip
if the
.Sq zip
tool is installed.
The optional
.Ar peer
argument can be used to specify the address or FQDN of the local gateway
which will be written into a text file
.Pa peer.txt
and included in the archives.
.It Xo
.Cm ca Ar name Cm certificate Ar host
.Cm install Op Ar path
.Xc
Install the private and public key for
.Ar host
into the active configuration or specified
.Ar path .
.It Xo
.Cm ca Ar name Cm certificate Ar host
.Cm revoke
.Xc
Revoke the certificate specified by
.Ar host
and generate a new Certificate Revocation List (CRL).
.It Xo
.Cm show Cm ca Ar name Cm certificates
.Op Ar host
.Xc
Display a listing of certificates associated with CA
.Ar name
or display certificate details if
.Ar host
is specified.
.It Xo
.Cm ca Ar name Cm key Ar host
.Cm create
.Xc
Create a private key for
.Ar host
if one does not already exist.
.It Xo
.Cm ca Ar name Cm key Ar host
.Cm install Op Ar path
.Xc
Install the private and public keys for
.Ar host
into the active configuration or specified
.Ar path .
.It Xo
.Cm ca Ar name Cm key Ar host
.Cm delete
.Xc
Delete the private key for
.Ar host .
.It Xo
.Cm ca Ar name Cm key Ar host
.Cm import
.Ar file
.Xc
Source the private key for
.Ar host
from the named
.Ar file .
.El
.Sh FILES
.Bl -tag -width "/var/run/iked.sockXX" -compact
.It Pa /etc/iked/
Active configuration.
.It Pa /etc/ssl/
Directory to store the CA files.
.It Pa /usr/share/iked/
If this optional directory exists,
.Nm
will include the contents with the
.Cm ca export
commands.
.It Pa /var/run/iked.sock
Default
.Ux Ns -domain
socket used for communication with
.Xr iked 8 .
.El
.Sh EXAMPLES
First create a new certificate authority:
.Bd -literal -offset indent
# ikectl ca vpn create
.Ed
.Pp
Now create the certificates for the VPN peers.
The specified hostname, either IP address or FQDN, will be saved in
the signed certificate and has to match the IKEv2 identity, or
.Ar srcid ,
of the peers:
.Bd -literal -offset indent
# ikectl ca vpn certificate 10.1.2.3 create
# ikectl ca vpn certificate 10.2.3.4 create
# ikectl ca vpn certificate 10.3.4.5 create
.Ed
.Pp
It is possible that the host that was used to create the CA is also
one of the VPN peers.
In this case you can install the peer and CA certificates locally:
.Bd -literal -offset indent
# ikectl ca vpn install
# ikectl ca vpn certificate 10.1.2.3 install
.Ed
.Pp
Now export the individual host key, the certificate and the CA
certificate to each other peer.
First run the
.Ic export
command to create tarballs that include the required files:
.Bd -literal -offset indent
# ikectl ca vpn certificate 10.2.3.4 export
# ikectl ca vpn certificate 10.3.4.5 export
.Ed
.Pp
These commands will produce two tarballs
.Em 10.2.3.4.tgz
and
.Em 10.3.4.5.tgz .
Copy these tarballs over to the appropriate peers and extract them
to the
.Pa /etc/iked/
directory:
.Bd -literal -offset indent
10.2.3.4# tar -C /etc/iked -xzpf 10.2.3.4.tgz
10.3.4.5# tar -C /etc/iked -xzpf 10.3.4.5.tgz
.Ed
.Pp
.Nm
will also create
.Sq zip
archives 10.2.3.4.zip and 10.3.4.5.zip
in addition to the tarballs if the zip tool is found in
.Pa /usr/local/bin/zip .
These archives can be exported to peers running Windows and will
include the certificates in a format that is supported by the OS.
.Sh SEE ALSO
.Xr packages 7 ,
.Xr iked 8 ,
.Xr ssl 8
.Sh HISTORY
The
.Nm
program first appeared in
.Ox 4.8 .
.Sh AUTHORS
The
.Nm
program was written by
.An Reyk Floeter Aq Mt reyk@openbsd.org
and
.An Jonathan Gray Aq Mt jsg@openbsd.org .
.Sh CAVEATS
For ease of use, the
.Ic ca
commands maintain all peers' private keys on the CA machine.
In contrast to a
.Sq real
CA, it does not support signing of public keys that have been imported
from peers that do not want to expose their private keys to the CA.


ikectl/ikectl.c
/*	$OpenBSD: ikectl.c,v 1.30 2022/12/04 11:54:31 tobhe Exp $	*/

/*
 * Copyright (c) 2007-2013 Reyk Floeter <reyk@openbsd.org>
 * Copyright (c) 2005 Claudio Jeker <claudio@openbsd.org>
 * Copyright (c) 2004, 2005 Esben Norby <norby@openbsd.org>
 * Copyright (c) 2003 Henning Brauer <henning@openbsd.org>
 *
 * Permission to use, copy, modify, and distribute this software for any
 * purpose with or without fee is hereby granted, provided that the above
 * copyright notice and this permission notice appear in all copies.
 *
 * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
 * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
 * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
 * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
 * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
 * ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
 * OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

#include <sys/types.h>
#include <sys/socket.h>
#include <sys/queue.h>
#include <sys/un.h>
#include <sys/tree.h>

#include <err.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <event.h>

#include "iked.h"
#include "parser.h"

__dead void	 usage(void);

struct imsgname {
	int type;
	char *name;
	void (*func)(struct imsg *);
};

struct imsgname *monitor_lookup(uint8_t);
void		 monitor_id(struct imsg *);
int		 monitor(struct imsg *);

int		 show_string(struct imsg *);
int		 show_stats(struct imsg *, int);

int		 ca_opt(struct parse_result *);

struct imsgname imsgs[] = {
	{ IMSG_CTL_OK,			"ok",			NULL },
	{ IMSG_CTL_FAIL,		"fail",			NULL },
	{ IMSG_CTL_VERBOSE,		"verbose",		NULL },
	{ IMSG_CTL_RELOAD,		"reload",		NULL },
	{ IMSG_CTL_RESET,		"reset",		NULL },
	{ IMSG_CTL_SHOW_SA,		"show sa",		NULL },
	{ IMSG_CTL_SHOW_CERTSTORE,	"show certstore",	NULL },
	{ 0,				NULL,			NULL }

};
struct imsgname imsgunknown = {
	-1,				"<unknown>",		NULL
};

struct imsgbuf	*ibuf;

__dead void
usage(void)
{
	extern char *__progname;

	fprintf(stderr, "usage: %s [-q] [-s socket] command [arg ...]\n",
	    __progname);
	exit(1);
}

int
ca_opt(struct parse_result *res)
{
	struct ca	*ca;
	size_t		 len;
	char		*p;

	ca = ca_setup(res->caname, (res->action == CA_CREATE),
	    res->quiet, res->pass);
	if (ca == NULL)
		errx(1, "ca_setup failed");

	/* assume paths are relative to /etc if not absolute */
	if (res->path && (res->path[0] != '.') && (res->path[0] != '/')) {
		len = 5 + strlen(res->path) + 1;
		if ((p = malloc(len)) == NULL)
			err(1, "malloc");
		snprintf(p, len, "/etc/%s", res->path);
		free(res->path);
		res->path = p;
	}

	switch (res->action) {
	case CA_CREATE:
		ca_create(ca);
		break;
	case CA_DELETE:
		ca_delete(ca);
		break;
	case CA_INSTALL:
		ca_install(ca, res->path);
		break;
	case CA_EXPORT:
		ca_export(ca, NULL, res->peer, res->pass);
		break;
	case CA_CERT_CREATE:
	case CA_SERVER:
	case CA_CLIENT:
	case CA_OCSP:
		ca_certificate(ca, res->host, res->htype, res->action);
		break;
	case CA_CERT_DELETE:
		ca_delkey(ca, res->host);
		break;
	case CA_CERT_INSTALL:
		ca_cert_install(ca, res->host, res->path);
		break;
	case CA_CERT_EXPORT:
		ca_export(ca, res->host, res->peer, res->pass);
		break;
	case CA_CERT_REVOKE:
		ca_revoke(ca, res->host);
		break;
	case SHOW_CA_CERTIFICATES:
		ca_show_certs(ca, res->host);
		break;
	case CA_KEY_CREATE:
		ca_key_create(ca, res->host);
		break;
	case CA_KEY_DELETE:
		ca_key_delete(ca, res->host);
		break;
	case CA_KEY_INSTALL:
		ca_key_install(ca, res->host, res->path);
		break;
	case CA_KEY_IMPORT:
		ca_key_import(ca, res->host, res->path);
		break;
	default:
		break;
	}

	return (0);
}

int
main(int argc, char *argv[])
{
	struct sockaddr_un	 s_un;
	struct parse_result	*res;
	struct imsg		 imsg;
	int			 ctl_sock;
	int			 done = 1;
	int			 n;
	int			 ch;
	int			 v = 0;
	int			 quiet = 0;
	const char		*sock = IKED_SOCKET;

	while ((ch = getopt(argc, argv, "qs:")) != -1) {
		switch (ch) {
		case 'q':
			quiet = 1;
			break;
		case 's':
			sock = optarg;
			break;
		default:
			usage();
			/* NOTREACHED */
		}
	}
	argc -= optind;
	argv += optind;

	/* parse options */
	if ((res = parse(argc, argv)) == NULL)
		exit(1);

	res->quiet = quiet;

	switch (res->action) {
	case CA_CREATE:
	case CA_DELETE:
	case CA_INSTALL:
	case CA_EXPORT:
	case CA_CERT_CREATE:
	case CA_CLIENT:
	case CA_SERVER:
	case CA_OCSP:
	case CA_CERT_DELETE:
	case CA_CERT_INSTALL:
	case CA_CERT_EXPORT:
	case CA_CERT_REVOKE:
	case SHOW_CA:
	case SHOW_CA_CERTIFICATES:
	case CA_KEY_CREATE:
	case CA_KEY_DELETE:
	case CA_KEY_INSTALL:
	case CA_KEY_IMPORT:
		if (pledge("stdio proc exec rpath wpath cpath fattr tty", NULL)
		    == -1)
			err(1, "pledge");
		ca_opt(res);
		break;
	case NONE:
		usage();
		break;
	default:
		goto connect;
	}

	return (0);

 connect:
	/* connect to iked control socket */
	if ((ctl_sock = socket(AF_UNIX, SOCK_STREAM, 0)) == -1)
		err(1, "socket");

	bzero(&s_un, sizeof(s_un));
	s_un.sun_family = AF_UNIX;
	strlcpy(s_un.sun_path, sock, sizeof(s_un.sun_path));
 reconnect:
	if (connect(ctl_sock, (struct sockaddr *)&s_un, sizeof(s_un)) == -1) {
		/* Keep retrying if running in monitor mode */
		if (res->action == MONITOR &&
		    (errno == ENOENT || errno == ECONNREFUSED)) {
			usleep(100);
			goto reconnect;
		}
		err(1, "connect: %s", sock);
	}

	if (pledge("stdio", NULL) == -1)
		err(1, "pledge");

	if (res->ibuf != NULL)
		ibuf = res->ibuf;
	else
		if ((ibuf = malloc(sizeof(struct imsgbuf))) == NULL)
			err(1, "malloc");
	imsg_init(ibuf, ctl_sock);

	/* process user request */
	switch (res->action) {
	case RESETALL:
		v = RESET_ALL;
		break;
	case RESETCA:
		v = RESET_CA;
		break;
	case RESETPOLICY:
		v = RESET_POLICY;
		break;
	case RESETSA:
		v = RESET_SA;
		break;
	case RESETUSER:
		v = RESET_USER;
		break;
	case LOG_VERBOSE:
		v = 2;
		break;
	case LOG_BRIEF:
	default:
		v = 0;
		break;
	}

	switch (res->action) {
	case NONE:
		usage();
		/* NOTREACHED */
		break;
	case RESETALL:
	case RESETCA:
	case RESETPOLICY:
	case RESETSA:
	case RESETUSER:
		imsg_compose(ibuf, IMSG_CTL_RESET, 0, 0, -1, &v, sizeof(v));
		printf("reset request sent.\n");
		break;
	case LOAD:
		imsg_compose(ibuf, IMSG_CTL_RELOAD, 0, 0, -1,
		    res->path, strlen(res->path));
		break;
	case RESET_ID:
		imsg_compose(ibuf, IMSG_CTL_RESET_ID, 0, 0, -1,
		    res->id, strlen(res->id));
		break;
	case SHOW_SA:
		imsg_compose(ibuf, IMSG_CTL_SHOW_SA, 0, 0, -1, NULL, 0);
		done = 0;
		break;
	case SHOW_STATS:
		imsg_compose(ibuf, IMSG_CTL_SHOW_STATS, 0, 0, -1, NULL, 0);
		done = 0;
		break;
	case SHOW_CERTSTORE:
		imsg_compose(ibuf, IMSG_CTL_SHOW_CERTSTORE, 0, 0, -1, NULL, 0);
		done = 0;
		break;
	case RELOAD:
		imsg_compose(ibuf, IMSG_CTL_RELOAD, 0, 0, -1, NULL, 0);
		break;
	case MONITOR:
		imsg_compose(ibuf, IMSG_CTL_NOTIFY, 0, 0, -1, NULL, 0);
		done = 0;
		break;
	case COUPLE:
		imsg_compose(ibuf, IMSG_CTL_COUPLE, 0, 0, -1, NULL, 0);
		break;
	case DECOUPLE:
		imsg_compose(ibuf, IMSG_CTL_DECOUPLE, 0, 0, -1, NULL, 0);
		break;
	case ACTIVE:
		imsg_compose(ibuf, IMSG_CTL_ACTIVE, 0, 0, -1, NULL, 0);
		break;
	case PASSIVE:
		imsg_compose(ibuf, IMSG_CTL_PASSIVE, 0, 0, -1, NULL, 0);
		break;
	case LOG_VERBOSE:
	case LOG_BRIEF:
		imsg_compose(ibuf, IMSG_CTL_VERBOSE, 0, 0, -1, &v, sizeof(v));
		printf("logging request sent.\n");
		break;
	default:
		break;
	}

	while (ibuf->w.queued)
		if (msgbuf_write(&ibuf->w) <= 0 && errno != EAGAIN)
			err(1, "write error");

	while (!done) {
		if ((n = imsg_read(ibuf)) == -1 && errno != EAGAIN)
			errx(1, "imsg_read error");
		if (n == 0)
			errx(1, "pipe closed");

		while (!done) {
			if ((n = imsg_get(ibuf, &imsg)) == -1)
				errx(1, "imsg_get error");
			if (n == 0)
				break;
			switch (res->action) {
			case MONITOR:
				done = monitor(&imsg);
				break;
			case SHOW_STATS:
				done = show_stats(&imsg, quiet);
				break;
			case SHOW_SA:
			case SHOW_CERTSTORE:
				done = show_string(&imsg);
				break;
			default:
				break;
			}
			imsg_free(&imsg);
		}
	}
	close(ctl_sock);
	free(ibuf);

	return (0);
}

struct imsgname *
monitor_lookup(uint8_t type)
{
	int i;

	for (i = 0; imsgs[i].name != NULL; i++)
		if (imsgs[i].type == type)
			return (&imsgs[i]);
	return (&imsgunknown);
}

int
monitor(struct imsg *imsg)
{
	time_t			 now;
	int			 done = 0;
	struct imsgname		*imn;

	now = time(NULL);

	imn = monitor_lookup(imsg->hdr.type);
	printf("%s: imsg type %u len %u peerid %u pid %d\n", imn->name,
	    imsg->hdr.type, imsg->hdr.len, imsg->hdr.peerid, imsg->hdr.pid);
	printf("\ttimestamp: %lld, %s", (long long)now, ctime(&now));
	if (imn->type == -1)
		done = 1;
	if (imn->func != NULL)
		(*imn->func)(imsg);

	return (done);
}

int
show_string(struct imsg *imsg)
{
	int	done = 0;

	switch (imsg->hdr.type) {
	case IMSG_CTL_SHOW_SA:
	case IMSG_CTL_SHOW_CERTSTORE:
		break;
	default:
		return (done);
	}
	if (IMSG_DATA_SIZE(imsg) > 0)
		printf("%s", (char *)imsg->data);
	else
		done = 1;

	return (done);
}

static char *
plural(uint64_t n)
{
        return (n != 1 ? "s" : "");
}

/*
 * Dump IKE statistics structure.
 */
int
show_stats(struct imsg *imsg, int quiet)
{
	struct iked_stats *stat;
	int		 done = 1;

	if (IMSG_DATA_SIZE(imsg) != sizeof(*stat))
		return (done);
	stat = imsg->data;
	printf("ike:\n");
#define p(f, m) if (stat->f || !quiet) \
	printf(m, (unsigned long long)stat->f, plural(stat->f))

	p(ikes_sa_created, "\t%llu IKE SA%s created\n");
	p(ikes_sa_established_total, "\t%llu IKE SA%s established\n");
	p(ikes_sa_established_current, "\t%llu IKE SA%s currently established\n");
	p(ikes_sa_established_failures, "\t%llu IKE SA%s failed to establish\n");
	p(ikes_sa_proposals_negotiate_failures, "\t%llu failed proposal negotiation%s\n");
	p(ikes_sa_rekeyed, "\t%llu IKE SA%s rekeyed\n");
	p(ikes_sa_removed, "\t%llu IKE SA%s removed\n");
	p(ikes_csa_created, "\t%llu Child SA%s created\n");
	p(ikes_csa_removed, "\t%llu Child SA%s removed\n");
	p(ikes_msg_sent, "\t%llu message%s sent\n");
	p(ikes_msg_send_failures, "\t%llu message%s could not be sent\n");
	p(ikes_msg_rcvd, "\t%llu message%s received\n");
	p(ikes_msg_rcvd_dropped, "\t%llu message%s dropped\n");
	p(ikes_msg_rcvd_busy, "\t%llu request%s dropped, response being worked on\n");
	p(ikes_retransmit_response, "\t%llu response%s retransmitted\n");
	p(ikes_retransmit_request, "\t%llu request%s retransmitted\n");
	p(ikes_retransmit_limit, "\t%llu request%s timed out\n");
	p(ikes_frag_sent, "\t%llu fragment%s sent\n");
	p(ikes_frag_send_failures, "\t%llu fragment%s could not be sent\n");
	p(ikes_frag_rcvd, "\t%llu fragment%s received\n");
	p(ikes_frag_rcvd_drop, "\t%llu fragment%s dropped\n");
	p(ikes_frag_reass_ok, "\t%llu fragment%s reassembled\n");
	p(ikes_frag_reass_drop, "\t%llu fragment%s could not be reassembled\n");
	p(ikes_update_addresses_sent, "\t%llu update addresses request%s sent\n");
	p(ikes_dpd_sent, "\t%llu dpd request%s sent\n");
	p(ikes_keepalive_sent, "\t%llu keepalive message%s sent\n");
#undef p
	return (done);
}
